### PR TITLE
DBZ-4667 Fix BEGIN/COMMIT handling when no VGTID was received

### PR DESCRIPTION
**Pull request by shichao-an**
Transaction ID in TransactionalMessage cannot be null. Therefore, when the decoder is handling BEGIN and COMMIT events and there is no VGTID event received (this can happen in VStream where the VStream Response consists of only BEGIN and COMMIT events), we do not process it.

This change has been tested in our corporate environment.

This fixes https://issues.redhat.com/browse/DBZ-4667

cc @y5w @keweishang @gunnarmorling

### CODE REVIEW

**Review by jeffchao — APPROVED** [review]
Looks good. Nice find!

**Review by y5w — COMMENTED** [review]
Good find. 

I think the reason we see BEGIN/COMMIT events with empty gtid is because Vitess only emits BEGIN/COMMIT without gtid if we are writing to internal tables (e.g., `_vt.*` tables). For our setup, the heartbeat writes will generate events you described in the ticket.

An idea for the testing setup: the existing test cases have decent coverages already, but I think we can add an integration test for this case: 
- create a dummy table e.g., `_vt.dummy`
- do an insert on the normal table (w1)
- do an insert on the `_vt.dummy` (w2)
- assert we only see event from w1 without error

**Comment by y5w** [comment]
> Good find.
> 
> I think the reason we see BEGIN/COMMIT events with empty gtid is because Vitess only emits BEGIN/COMMIT without gtid if we are writing to internal tables (e.g., `_vt.*` tables). For our setup, the heartbeat writes will generate events you described in the ticket.
> 
> An idea for the testing setup: the existing test cases have decent coverages already, but I think we can add an integration test for this case:
> 
> * create a dummy table e.g., `_vt.dummy`
> * do an insert on the normal table (w1)
> * do an insert on the `_vt.dummy` (w2)
> * assert we only see event from w1 without error

Not a blocking comment btw. If we want to get in before the next release, feel free do it as a follow up.

**Comment by shichao-an** [comment]
> Good find.
> 
> I think the reason we see BEGIN/COMMIT events with empty gtid is because Vitess only emits BEGIN/COMMIT without gtid if we are writing to internal tables (e.g., `_vt.*` tables). For our setup, the heartbeat writes will generate events you described in the ticket.

I'm not sure what triggers such events, but I'm pretty sure it's not heartbeat writes. I remember we had vttablet hearbeats enabled when we first set up the prototype environment, and the transaction_metadata topic was being bombarded by heartbeat events, and we never encountered this issue. About two months ago the team disabled vttablet lag throttler which disabled heartbeats as a side-effect.

I just manually tested the heartbeats as follows and couldn't reproduce that.

In vttablet mysql:
```
mysql> use _vt;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A
Database changed
mysql> UPDATE heartbeat set ts = 1643917449794173639;
Query OK, 1 row affected (0.00 sec)
Rows matched: 1  Changed: 1  Warnings: 0
```

VStream command:
```
$ grpcurl -d '{ "vgtid": { "shard_gtids": [ { "keyspace": "test_keyspace", "shard": "-80", "gtid": "current" } ] } }' -plaintext xxxx:1234 vtgateservice.Vitess.VStream
{
  "events": [
    {
      "type": "VGTID",
      "vgtid": {
        "shardGtids": [
          {
            "keyspace": "test_keyspace",
            "shard": "-80",
            "gtid": "MySQL56/2582b27c-7e73-4cd2-8640-028069dc2704:1-2575467:3343958-3344114"
          }
        ]
      }
    },
    {
      "type": "OTHER"
    }
  ]
}
{
  "events": [
    {
      "type": "BEGIN",
      "timestamp": "1643917461",
      "currentTime": "1643917461158158827"
    },
    {
      "type": "VGTID",
      "vgtid": {
        "shardGtids": [
          {
            "keyspace": "test_keyspace",
            "shard": "-80",
            "gtid": "MySQL56/2582b27c-7e73-4cd2-8640-028069dc2704:1-2575468:3343958-3344114"
          }
        ]
      }
    },
    {
      "type": "COMMIT",
      "timestamp": "1643917461",
      "currentTime": "1643917461158209298"
    }
  ]
}
```

File: src/main/java/io/debezium/connector/vitess/connection/VStreamOutputMessageDecoder.java [hunk 1]
```diff
@@ -110,6 +111,11 @@ private void handleBeginMessage(Binlogdata.VEvent vEvent, ReplicationMessageProc
         if (newVgtid != null) {
             this.transactionId = newVgtid.toString();
         }
+        // Transaction ID must not be null in TransactionalMessage.
+        if (this.transactionId == null) {
```
[the remark this answers is not in the document]
Comment (y5w) [reply src/main/java/io/debezium/connector/vitess/connection/VStreamOutputMessageDecoder.java:115]: Yeah, I was thinking about a check like
```
if (newVgtid.toString().isEmpty()) {
  return;
}
```

Or even do that check earlier: https://github.com/debezium/debezium-connector-vitess/blob/f3f8d6f13b66939222f848ac37fcb3c811593400/src/main/java/io/debezium/connector/vitess/connection/VitessReplicationConnection.java#L88

Basically, if a vstream response does not have gtid, I think it should be safe to skip.

Both approaches are doing the same thing, just doing a little fewer extra works and might also avoid some other corner cases introduced by events without gtid field

File: src/test/java/io/debezium/connector/vitess/connection/VStreamOutputMessageDecoderTest.java [hunk 2]
```diff
@@ -58,20 +62,49 @@ public void shouldProcessBeginEvent() throws Exception {
...
+
     @Test
     public void shouldProcessCommitEvent() throws Exception {
         // setup fixture
         Binlogdata.VEvent event = Binlogdata.VEvent.newBuilder()
                 .setType(Binlogdata.VEventType.COMMIT)
                 .setTimestamp(AnonymousValue.getLong())
                 .build();
+        Vgtid newVgtid = Vgtid.of(VgtidTest.VGTID_JSON);
+        decoder.setTransactionId(newVgtid.toString());
```
[the remark this answers is not in the document]
Comment (y5w) [reply src/test/java/io/debezium/connector/vitess/connection/VStreamOutputMessageDecoderTest.java:107]: I see, do we need to clear it out after this test? Otherwise all other test cases will use a decoder with txn id set

**Comment by y5w** [comment]
> > Good find.
> > I think the reason we see BEGIN/COMMIT events with empty gtid is because Vitess only emits BEGIN/COMMIT without gtid if we are writing to internal tables (e.g., `_vt.*` tables). For our setup, the heartbeat writes will generate events you described in the ticket.
> 
> I'm not sure what triggers such events, but I'm pretty sure it's not heartbeat writes. I remember we had vttablet hearbeats enabled when we first set up the prototype environment, and the transaction_metadata topic was being bombarded by heartbeat events, and we never encountered this issue. About two months ago the team disabled vttablet lag throttler which disabled heartbeats as a side-effect.
> 
> I just manually tested the heartbeats as follows and couldn't reproduce that.
> 
> In vttablet mysql:
> 
> ```
> mysql> use _vt;
> Reading table information for completion of table and column names
> You can turn off this feature to get a quicker startup with -A
> Database changed
> mysql> UPDATE heartbeat set ts = 1643917449794173639;
> Query OK, 1 row affected (0.00 sec)
> Rows matched: 1  Changed: 1  Warnings: 0
> ```
> 
> VStream command:
> 
> ```
> $ grpcurl -d '{ "vgtid": { "shard_gtids": [ { "keyspace": "test_keyspace", "shard": "-80", "gtid": "current" } ] } }' -plaintext xxxx:1234 vtgateservice.Vitess.VStream
> {
>   "events": [
>     {
>       "type": "VGTID",
>       "vgtid": {
>         "shardGtids": [
>           {
>             "keyspace": "test_keyspace",
>             "shard": "-80",
>             "gtid": "MySQL56/2582b27c-7e73-4cd2-8640-028069dc2704:1-2575467:3343958-3344114"
>           }
>         ]
>       }
>     },
>     {
>       "type": "OTHER"
>     }
>   ]
> }
> {
>   "events": [
>     {
>       "type": "BEGIN",
>       "timestamp": "1643917461",
>       "currentTime": "1643917461158158827"
>     },
>     {
>       "type": "VGTID",
>       "vgtid": {
>         "shardGtids": [
>           {
>             "keyspace": "test_keyspace",
>             "shard": "-80",
>             "gtid": "MySQL56/2582b27c-7e73-4cd2-8640-028069dc2704:1-2575468:3343958-3344114"
>           }
>         ]
>       }
>     },
>     {
>       "type": "COMMIT",
>       "timestamp": "1643917461",
>       "currentTime": "1643917461158209298"
>     }
>   ]
> }
> ```

Oh, I confused ROW_EVENT with VGTID_EVENT, my bad.

My point on the "internal" table event is we should filter them as well, or at very least add a flag to filter them out. Otherwise they will generate a lot meaningless events (think of our own heartbeat writer that is doing 1 write per sec per shard). I filed https://issues.redhat.com/browse/DBZ-4673 for that issue

**Review by y5w — APPROVED** [review]
LGTM, two minor comments.

File: src/test/java/io/debezium/connector/vitess/connection/VStreamOutputMessageDecoderTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (shichao-an) [reply src/test/java/io/debezium/connector/vitess/connection/VStreamOutputMessageDecoderTest.java:107]: No, there is `before` method called before each test function that creates a new decoder every time.

File: src/main/java/io/debezium/connector/vitess/connection/VStreamOutputMessageDecoder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (shichao-an) [reply src/main/java/io/debezium/connector/vitess/connection/VStreamOutputMessageDecoder.java:115]: Yes, the alternative approach would be checking `getVgtid(response)` and skip at that level, which was my first thought. Then, I read the existing code and understood the original intention of the code, so I made the changes in this PR (minimal change to the original intention). To summarize, it leverages the "cached" transaction id (vgtid) and also gave other events (e.g. other, ddl, field) a chance for processing that don't need transaction id.

**Comment by shichao-an** [comment]
> My point on the "internal" table event is we should filter them as well, or at very least add a flag to filter them out. Otherwise they will generate a lot meaningless events (think of our own heartbeat writer that is doing 1 write per sec per shard). I filed https://issues.redhat.com/browse/DBZ-4673 for that issue

Yes, I thought of this and didn't have a chance to dig further due to complications. These heartbeats events are also incrementing offsets via transaction events (in debezium core). This is something we actually want. However, I didn't figure out how we can commit offsets without sending actual transaction metadata events to Kafka. If we filter them out completely (without committing any offsets for them), we will have longer gaps between offsets, that means more overhead when being replayed, which is probably also fine considering low rate of the heartbeats (not much gaps either).

Alternative is that we simply don't filter them, but tune the heartbeat rate (default is 1 per second for lag throttler). The worst case is that these events coming out from each shard at QPS of 1 * N (N = number of shards), that means extra N qps on the transaction_metadata topic, so it's actually fine. Our downstream CDC Aggregator is already able to filter these empty transaction events (cc @jeffchao ) so they have no effect on them as well.

In any case we want to tune `offset.flush.interval.ms` in Kafka Connect. The default is 60 seconds, which is too small for constant load of traffic, making replay slower.

**Comment by y5w** [comment]
> If we filter them out completely (without committing any offsets for them), we will have longer gaps between offsets, that means more overhead when being replayed, which is probably also fine considering low rate of the heartbeats (not much gaps either).
> 
> Alternative is that we simply don't filter them, but tune the heartbeat rate (default is 1 per second for lag throttler). The worst case is that these events coming out from each shard at QPS of 1 * N (N = number of shards), that means extra N qps on the transaction_metadata topic, so it's actually fine.

My two cents are leaning towards filter them completely at the connector:
- Less data need to be shuffled around 
- Less data to consumed and filter in the aggregator 
- Tuning heartbeat interval has implication on the SLC on the Vitess data plane, especially given 1 RPS per shard overhead is not too bad right now
